In [2]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [3]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_df = pd.read_csv(
    Path("lending_data.csv")
)

# Review the DataFrame
display(lending_df.head())

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_df['loan_status']

# Separate the X variable, the features
x = lending_df.drop(columns="loan_status")

In [5]:
# Review the y variable Series
display(y.head())

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [6]:
# Review the X variable DataFrame
display(x.head())

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function

x_train, x_test, y_train, y_test = train_test_split(x,y, random_state=1, stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression (solver='lbfgs', max_iter=200, random_state=1)

# Fit the model using training data
classifier.fit(x_train, y_train)

print(f'{classifier.score(x_train,y_train)}')
print(f'{classifier.score(x_test,y_test)}')

0.9914878250103177
0.9924164259182832


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
prediction = classifier.predict(x_test)
results = pd.DataFrame({"Prediction": prediction, "Actual": y_test}).reset_index(drop=True)

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [11]:
# Generate a confusion matrix for the model
from sklearn.metrics import classification_report

confusion_matrix(y_test, prediction)

array([[18679,    80],
       [   67,   558]])

In [13]:
# Print the classification report for the model
from sklearn.metrics import classification_report

target_names = ["Healthy Loan", "High-Rick loan"]

print(classification_report(y_test, prediction, target_names=target_names))

                precision    recall  f1-score   support

  Healthy Loan       1.00      1.00      1.00     18759
High-Rick loan       0.87      0.89      0.88       625

      accuracy                           0.99     19384
     macro avg       0.94      0.94      0.94     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Healthy Loan: The model performs flawlessly when predicting Healthy Loan, with precision, recall, and F1-scores all at 1.00. This means the model is highly accurate in both identifying true positives (recall) and avoiding false positives (precision) for Healthy Loan. This is based on a large number of cases, with a support of 18,759, suggesting a strong and reliable performance for this class.
High-Rick loan: The model shows good but not perfect performance for High-Rick loan, with a precision of 0.87, indicating that 87% of the predicted high-risk loan are correctly classified. The recall is 0.89, meaning that 89% of all actual High-Rick loans were correctly identified by the model. The F1-score is 0.88, reflecting a balance between precision and recall. Given that the support for High-Rick loan is much smaller at 625, this slight drop in performance is expected due to the class imbalance.

---